hal/src/dmac/async_api.rs

```diff
@@ -96,6 +97,16 @@ impl Handler<DMAC> for InterruptHandler {
                     w.enable().clear_bit();
                     w.trigsrc().variant(TriggerSource::Disable)
                 });
+
+                while dmac.channel(channel).chctrla().read().enable().bit_is_set() {
+                    core::hint::spin_loop();
+                }
+
+                // Prevent the compiler from re-ordering read/write
+                // operations beyond this fence.
+                // (see https://docs.rust-embedded.org/embedonomicon/dma.html#compiler-misoptimizations)
+                atomic::fence(atomic::Ordering::Acquire); // ▼
```

Comment:
Is there another `atomic::fence` with `Release` ordering somewhere else? I thought fences like this were only meaningful when paired across two concurrent functions? I think it's already guaranteed by the language that the enable bit will read true before `wake()` is called. The reads are volatile, so they can't be removed, and the read is observed, so you can't reorder `wake()` before it.

Reply:
Yeah, this seems more or less unnecessary. I think the intention was to protect reordering reads/writes to the memory chunk the DMA is accessing, although we already have fences for that at a lower level.

That being said, we had a discussion a while ago about how `atomic::fence` isn't even enough to guarantee no reordering will take place... IIRC the verdict was that we should use an empty `asm!` black box instead.

Reply:
`Channel::_enable_private` has the corresponding `Release` fence + is called in the `TransferFuture::poll` implementation that this ISR interacts with

the